### ISSUE TITLE
saw: vecIdx (atOp[Nat]) out of bounds (4 out of 1)

### ISSUE THREAD

**Issue by weaversa** [GaloisInc/saw-script#97]
```
$ cat test.cry
pow : [32] -> [8] -> [32]
pow x e = exps@e
  where exps = [1] # [x * exp | exp <- exps]
$ saw
   ___  __ _ _ _ _
  / __|/ _' | | | |
  \__ \ (_| | | | |
  |___/\__,_|\_,_/ 

Loading module Cryptol
sawscript> :l test.cry
Loading module Main
sawscript> print {{ pow 2 4 }}
16
sawscript> prove z3 {{ pow 2 4 == 16 }}
saw: vecIdx (atOp[Nat]) out of bounds (4 out of 1)
```

**Comment by brianhuffman** [comment]
In that example, it appears that `exps` has type `[inf][32]`. So this probably indicates a bug in the SBV backend with either the append or indexing primitives for infinite streams.

**Comment by brianhuffman** [comment]
Here's what the definition of `pow` reduces to:

    sawscript> rewrite (cryptol_ss ()) (unfold_term ["pow"] {{ pow }})
    let { x0 = Prelude.Vec 32 Prelude.Bool;
          x1 = Prelude.Stream x0;
        }
     in \(x::x0) ->
          \(e::Prelude.Vec 8 Prelude.Bool) ->
            Prelude.bvStreamGet x0 8
              (Cryptol.fix x1
                 (\(exps::x1) ->
                    Prelude.streamAppend x0 1
                      [Prelude.bvNat 32 1]
                      (Prelude.streamMap x0 x0
                         (\(exp::x0) ->
                            Prelude.bvMul 32 x exp)
                         exps)))
              e

and here's the definition of `streamAppend` from the SAWCore prelude:

    streamAppend :: (a :: sort 0) -> (n :: Nat) -> Vec n a -> Stream a -> Stream a;
    streamAppend a n xs ys =
      MkStream a (\(i :: Nat) ->
        ite a (ltNat i n) (at n a xs i) (streamGet a ys (subNat i n)));

It's the call to `at` here that's causing the "out of bounds" error. The call is surrounded by an explicit bounds check, which is an effective guard for concrete evaluation. But apparently during symbolic execution both branches are always simulated.

The easy fix would be to redefine `at` to return some `zero` for out of bounds lookups. A better fix would be to make sure that `ite` is sufficiently lazy for symbolic execution. In this case `ltNat i n` should always be known statically for any list position, so we shouldn't have to evaluate the `then` branch when the guard is false.

**Comment by robdockins** [comment]
The symptoms of this issue are fixed by GaloisInc/saw-core@13b1af08c57292ba44a781bd94c2e6d3275c6534, which avoids simulating brances of 'ite' where we can concretely tell which branch to take.

However, I think we should also consider making 'at' and such return a defined value (zero, probably) for out-of-bounds lookups so that this sort of code will work even when the branch condition is symbolic.

**Comment by brianhuffman** [comment]
In Cryptol, it's easy to make the indexing `(@)` total by returning the `zero` for out-of-bounds accesses. But in SAWCore, `at` can be used with arbitrary element types, which won't necessarily have a zero value.

We might consider defining an `atWithDefault` SAWCore primitive that takes an extra parameter to use for out-of-bounds lookups. We could redefine `streamAppend` etc. using that instead.
